In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

In [2]:
taxi = pd.read_csv('clean_taxi.csv')

In [3]:
taxi.head(10)

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       1.0  2021-07-01 00:30:52   2021-07-01 00:35:36                  N   
1       2.0  2021-07-01 00:25:36   2021-07-01 01:01:31                  N   
2       2.0  2021-07-01 00:05:58   2021-07-01 00:12:00                  N   
3       2.0  2021-07-01 00:41:40   2021-07-01 00:47:23                  N   
4       2.0  2021-07-01 00:51:32   2021-07-01 00:58:46                  N   
5       1.0  2021-07-01 00:05:00   2021-07-01 00:11:50                  N   
6       2.0  2021-07-01 00:57:14   2021-07-01 01:27:43                  N   
7       2.0  2021-07-01 00:27:36   2021-07-01 00:32:35                  N   
8       2.0  2021-07-01 00:29:09   2021-07-01 00:34:18                  N   
9       2.0  2021-07-01 00:41:33   2021-07-01 00:49:24                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            74           168              1.0           1.20   
1         1.0           116           265              2.0          13.69   
2         1.0            97            33              1.0           0.95   
3         1.0            74            42              1.0           1.24   
4         1.0            42           244              1.0           1.10   
5         1.0            24           239              1.0           1.90   
6         1.0            75           243              1.0           0.00   
7         1.0            82            82              1.0           0.66   
8         1.0            74            42              1.0           1.72   
9         1.0            41            42              1.0           1.37   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0          6.0   0.50      0.5        0.00           0.0   
1         42.0   0.50      0.5        0.00           0.0   
2          6.5   0.50      0.5        2.34           0.0   
3          6.5   0.50      0.5        0.00           0.0   
4          7.0   0.50      0.5        0.00           0.0   
5          8.0   3.25      0.5        3.00           0.0   
6         17.5   0.50      0.5        0.00           0.0   
7          5.0   0.50      0.5        0.00           0.0   
8          7.0   0.50      0.5        2.08           0.0   
9          7.5   0.50      0.5        0.00           0.0   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          7.30           2.0        1.0   
1                    0.3         43.30           2.0        1.0   
2                    0.3         10.14           1.0        1.0   
3                    0.3          7.80           2.0        1.0   
4                    0.3          8.30           2.0        1.0   
5                    0.3         15.05           1.0        1.0   
6                    0.3         18.80           2.0        1.0   
7                    0.3          6.30           2.0        1.0   
8                    0.3         10.38           1.0        1.0   
9                    0.3          8.80           2.0        1.0   

   congestion_surcharge  
0                  0.00  
1                  0.00  
2                  0.00  
3                  0.00  
4                  0.00  
5                  2.75  
6                  0.00  
7                  0.00  
8                  0.00  
9                  0.00

In [4]:
taxi.describe()

VendorID    RatecodeID  PULocationID  DOLocationID  \
count  51173.000000  51173.000000  51173.000000  51173.000000   
mean       1.851113      1.159244     98.786528    135.229613   
std        0.355981      0.773260     64.719493     77.737705   
min        1.000000      1.000000      3.000000      1.000000   
25%        2.000000      1.000000     52.000000     74.000000   
50%        2.000000      1.000000     75.000000    135.000000   
75%        2.000000      1.000000    129.000000    213.000000   
max        2.000000      5.000000    265.000000    265.000000   

       passenger_count  trip_distance   fare_amount         extra  \
count     51173.000000   51173.000000  51173.000000  51173.000000   
mean          1.307858       3.494247     15.034696      0.360429   
std           0.984362       4.620208     14.545213      0.624490   
min           0.000000       0.000000      0.000000      0.000000   
25%           1.000000       1.100000      7.000000      0.000000   
50%           1.000000       1.990000     10.500000      0.000000   
75%           1.000000       3.860000     17.000000      0.500000   
max          32.000000     109.870000    480.000000      4.500000   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  51173.000000  51173.000000  51173.000000           51173.000000   
mean       0.480243      1.379782      0.342628               0.297966   
std        0.097407      2.604697      1.514054               0.024620   
min        0.000000      0.000000      0.000000               0.000000   
25%        0.500000      0.000000      0.000000               0.300000   
50%        0.500000      0.000000      0.000000               0.300000   
75%        0.500000      2.160000      0.000000               0.300000   
max        0.500000     87.710000     30.050000               0.300000   

       total_amount  payment_type     trip_type  congestion_surcharge  
count  51173.000000  51173.000000  51173.000000          51173.000000  
mean      18.454531      1.421726      1.034393              0.643137  
std       16.201376      0.511146      0.182239              1.164041  
min        0.000000      1.000000      1.000000              0.000000  
25%        9.300000      1.000000      1.000000              0.000000  
50%       13.560000      1.000000      1.000000              0.000000  
75%       21.300000      2.000000      1.000000              0.000000  
max      480.310000      5.000000      2.000000              2.750000

In [5]:
taxi = taxi.drop(['lpep_pickup_datetime', 'lpep_dropoff_datetime'],1)

/var/folders/x1/k6x2x2c9677_n0slz_pq1p5r0000gn/T/ipykernel_12344/2963341786.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  taxi = taxi.drop(['lpep_pickup_datetime', 'lpep_dropoff_datetime'],1)


In [6]:
taxi.groupby(['payment_type','tip_amount']).mean()

VendorID  RatecodeID  PULocationID  DOLocationID  \
payment_type tip_amount                                                     
1.0          0.00        1.777228    1.223974    114.719097    130.563143   
             0.01        1.913043    1.434783    116.217391    147.782609   
             0.02        1.913043    1.478261     86.913043    132.869565   
             0.03        1.900000    1.000000     72.400000    105.900000   
             0.04        1.833333    1.666667    125.666667     77.666667   
...                           ...         ...           ...           ...   
3.0          0.76        2.000000    1.000000    155.800000    193.000000   
             1.01        2.000000    1.000000    193.000000    193.000000   
             1.14        2.000000    1.000000    193.000000    193.000000   
4.0          0.00        1.431818    1.363636     94.863636     98.772727   
5.0          0.00        1.000000    1.000000     82.000000     48.000000   

                         passenger_count  trip_distance  fare_amount  \
payment_type tip_amount                                                
1.0          0.00               1.219564       7.064267    30.284053   
             0.01               1.130435       2.029130    16.816304   
             0.02               1.086957       2.959565    13.804348   
             0.03               2.200000       2.104000    12.150000   
             0.04               1.166667       3.985000    14.500000   
...                                  ...            ...          ...   
3.0          0.76               1.000000       0.000000     3.000000   
             1.01               1.000000       0.000000     2.500000   
             1.14               1.000000       0.000000     3.000000   
4.0          0.00               1.159091       0.826591     7.693182   
5.0          0.00               1.000000       8.100000    35.000000   

                            extra   mta_tax  tolls_amount  \
payment_type tip_amount                                     
1.0          0.00        0.245189  0.466858      1.346245   
             0.01        0.296196  0.445652      0.103804   
             0.02        0.250000  0.456522      0.000000   
             0.03        0.400000  0.500000      0.000000   
             0.04        0.166667  0.416667      0.000000   
...                           ...       ...           ...   
3.0          0.76        0.000000  0.500000      0.000000   
             1.01        0.000000  0.500000      0.000000   
             1.14        0.000000  0.500000      0.000000   
4.0          0.00        0.323864  0.454545      0.000000   
5.0          0.00        3.750000  0.500000      0.000000   

                         improvement_surcharge  total_amount  trip_type  \
payment_type tip_amount                                                   
1.0          0.00                     0.296111     32.801499   1.051851   
             0.01                     0.300000     18.270870   1.097826   
             0.02                     0.300000     15.309130   1.086957   
             0.03                     0.300000     13.655000   1.000000   
             0.04                     0.300000     15.423333   1.166667   
...                                        ...           ...        ...   
3.0          0.76                     0.300000      4.560000   1.000000   
             1.01                     0.300000      4.310000   1.000000   
             1.14                     0.300000      4.940000   1.000000   
4.0          0.00                     0.293182      8.764773   1.090909   
5.0          0.00                     0.300000     39.550000   1.000000   

                         congestion_surcharge  
payment_type tip_amount                        
1.0          0.00                    0.187057  
             0.01                    0.328804  
             0.02                    0.597826  
             0.03                    0.275000  
             0.04           

In [7]:
taxi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51173 entries, 0 to 51172
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               51173 non-null  float64
 1   store_and_fwd_flag     51173 non-null  object 
 2   RatecodeID             51173 non-null  float64
 3   PULocationID           51173 non-null  int64  
 4   DOLocationID           51173 non-null  int64  
 5   passenger_count        51173 non-null  float64
 6   trip_distance          51173 non-null  float64
 7   fare_amount            51173 non-null  float64
 8   extra                  51173 non-null  float64
 9   mta_tax                51173 non-null  float64
 10  tip_amount             51173 non-null  float64
 11  tolls_amount           51173 non-null  float64
 12  improvement_surcharge  51173 non-null  float64
 13  total_amount           51173 non-null  float64
 14  payment_type           51173 non-null  float64
 15  tr

In [8]:
le = LabelEncoder()

In [9]:
taxi['store_and_fwd_flag'] = le.fit_transform(taxi['store_and_fwd_flag'])
taxi.head()

VendorID  store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  \
0       1.0                   0         1.0            74           168   
1       2.0                   0         1.0           116           265   
2       2.0                   0         1.0            97            33   
3       2.0                   0         1.0            74            42   
4       2.0                   0         1.0            42           244   

   passenger_count  trip_distance  fare_amount  extra  mta_tax  tip_amount  \
0              1.0           1.20          6.0    0.5      0.5        0.00   
1              2.0          13.69         42.0    0.5      0.5        0.00   
2              1.0           0.95          6.5    0.5      0.5        2.34   
3              1.0           1.24          6.5    0.5      0.5        0.00   
4              1.0           1.10          7.0    0.5      0.5        0.00   

   tolls_amount  improvement_surcharge  total_amount  payment_type  trip_type  \
0           0.0                    0.3          7.30           2.0        1.0   
1           0.0                    0.3         43.30           2.0        1.0   
2           0.0                    0.3         10.14           1.0        1.0   
3           0.0                    0.3          7.80           2.0        1.0   
4           0.0                    0.3          8.30           2.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0

In [10]:
X = taxi.drop('total_amount', axis=1)
y = taxi['total_amount']

In [11]:
X = pd.get_dummies(data=X)
y = pd.get_dummies(data=y)

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, random_state=123)

In [13]:
X_train.dtypes

VendorID                 float64
store_and_fwd_flag         int64
RatecodeID               float64
PULocationID               int64
DOLocationID               int64
passenger_count          float64
trip_distance            float64
fare_amount              float64
extra                    float64
mta_tax                  float64
tip_amount               float64
tolls_amount             float64
improvement_surcharge    float64
payment_type             float64
trip_type                float64
congestion_surcharge     float64
dtype: object

In [14]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [15]:
scaler.fit_transform(X_train)

array([[ 0.41768395, -0.06347265,  4.9902647 , ..., -0.82289816,
         5.31109806, -0.55098956],
       [ 0.41768395, -0.06347265, -0.2050497 , ..., -0.82289816,
        -0.18828498, -0.55098956],
       [-2.39415472, -0.06347265, -0.2050497 , ...,  1.1297424 ,
        -0.18828498, -0.55098956],
       ...,
       [-2.39415472, -0.06347265, -0.2050497 , ..., -0.82289816,
        -0.18828498,  1.81496016],
       [ 0.41768395, -0.06347265, -0.2050497 , ...,  1.1297424 ,
        -0.18828498, -0.55098956],
       [ 0.41768395, -0.06347265, -0.2050497 , ...,  1.1297424 ,
        -0.18828498, -0.55098956]])

In [16]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn import metrics

In [17]:
lr = LinearRegression()
dt = DecisionTreeRegressor()
knn = KNeighborsRegressor()

In [18]:
lr.fit(X_train, y_train)
lin_pred = lr.predict(X_test)

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, lin_pred)))
print('R-Squared:', metrics.r2_score(y_test, lin_pred))

Root Mean Squared Error: 0.0206585877145536
R-Squared: 0.0015885008810073004


In [19]:
dt.fit(X_train, y_train)
dt_pred = dt.predict(X_test)

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, dt_pred)))
print('R-Squared:', metrics.r2_score(y_test, dt_pred))

Root Mean Squared Error: 0.01063345809031152
R-Squared: 0.48649501602870504


In [20]:
knn.fit(X_train, y_train)
knn_pred = knn.predict(X_test)

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, knn_pred)))
print('R-Squared:', metrics.r2_score(y_test, knn_pred))

Root Mean Squared Error: 0.020242573764615194
R-Squared: 0.0594381179451874


The most accurate model is the Decision Tree Regressor with the RMSE value of ~0.01 and a R2 value of ~0.49